src/ip/mod.rs

```diff
@@ -1,3 +1,11 @@
+#[derive(Debug, PartialEq)]
+struct NetworkAddr<T>(T);

```

Comment:
The only reason I derive from `Debug` and `PartialEq` is to support the `assert_eq!` macro, in the tests (at the bottom of `mod.rs`)

```rust
#[test]
fn test_network_values() {
    ...
    assert_eq!((NetworkAddr(network), BroadcastAddr(broadcast)), 
        calc_network_values("192.168.0.15", mask).unwrap());
}
```

The macro requires these traits to check equality and to `fmt` output in case the assert fails. Otherwise, these traits aren't necessary. Actually, the `derive` attributes can be conditionally compiled out by replacing those lines with `#[cfg_attr(test, derive(Debug, PartialEq))]`, so that they're only included when building / running the `test` config (E.g. running `cargo test`).

Please forgive the large PR. What you're seeing the result of me dissecting the code to try and find the bug. I added `NetworkAddr` and `BroadcastAddr` so that the compiler would type-check in case I was mixing up the two somewhere. Even when I changed `make_ip_from_mask` to bit shift, I couldn't see why it worked!  *- turns out the division was off-by-one; dividing by `0x100` for the 3rd octet would work, for example)*.